File: Cousins_in_Binary_Tree_BFS.py
# Created by Aashish Adhikari at 8:34 PM 1/21/2021
'''
Time Complexity:
O(n) as we reach all the nodes.

Space Complexity:
O(n/2) as we are storing all the nodes of the same generation in the queue and the last level has n/2 nodes in a complete binary tree.
'''

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isCousins(self, root, x, y):
        """
        :type root: TreeNode
        :type x: int
        :type y: int
        :rtype: bool
        """


        from collections import deque

        level = deque()
        level.append(root)


        while len(level) != 0:

            leng = len(level)

            # set for this level
            cousins = set()

            for idx in range(0, leng):
                node = level[0]

                if node.left is not None and node.right is not None:
                    if node.left.val == x and node.right.val == y or node.left.val == y and node.right.val == x:
                        # siblings, not cousins
                        return False
                    else:
                        cousins.add(node.left.val)
                        cousins.add(node.right.val)

                        level.append(node.left)
                        level.append(node.right)
                elif node.left is not None:
                    level.append(node.left)
                    cousins.add(node.left.val)
                elif node.right is not None:
                    level.append(node.right)
                    cousins.add(node.right.val)

                else:
                    pass

                level.popleft()


            if x in cousins and y in cousins:
                return True
            elif x in cousins:
                return False
            elif y in cousins:
                return False
            else:
                pass

        return False

File: cousinsInBinaryTreeBFS.java
// Time Complexity: O(n) where n is the number of nodes.
// Space Complexity: O(n) as Queue is going to occupy the maximum space in worst case by holding n/2 elements.
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        List<TreeNode> target = new ArrayList<>();
        List<Integer> height = new ArrayList<>();
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int depth = 0;
        while(!q.isEmpty()){
            int size = q.size();
            if(target.size() >= 1) break;        // break if in previous iteration of level we found only one element
            for(int i=0; i< size; i++){
                TreeNode curr = q.poll();
                if(curr.left != null){
                    q.add(curr.left);
                    if(curr.left.val == x || curr.left.val == y){
                        target.add(curr);
                        height.add(depth);
                    }
                }
                if(curr.right != null){
                    q.add(curr.right);
                    if(curr.right.val == x || curr.right.val == y){
                        target.add(curr);
                        height.add(depth);
                    }
                }
            }
            depth++;
        }
        // Confirm if we found both element and they adhere to all given conditions
        if(target.size() == 2 && target.get(0) != target.get(1) && height.get(0) == height.get(1)) return true;
        return false;
    }
}